#' @title Applying an fixed moving window
#'
#' @usage FixedMovingWindow(data, gencode, window_size = 151)
#'
#' @param data data.frame; Normalized read counts with cells in columns and genes
#'  (hgnc symbol) in rows.
#' @param gencode; gene annotation file
#' @param window_size; size of the used window
#'
#' @return data frame with relative smoothed CNA scores from all input cells
#'
#' @description normalized expression values are averaged across neighboring genes
#'
#' @export
#'
FixedMovingWindow <- function(data, gencode, window_size = 151){
    gCNA <- genome_sort(data, gencode)

    message("Centering normalized expression profiles per cell")
    gCNA <- apply(gCNA, 2, function(x) x-mean(x))

    window <-rep(window_size,ncol(data)) # set a fixed moving window

    cat("Window size:\n")
    print(window_size)
    message("Smoothing gene-specific CNAs.")
    rCNA <- cnvperchr.weight(data = gCNA,
                             chromosomeorder = chromosomeorder,
                             windowlength = window)
    return(rCNA)
}

## Function to sort data frames according to the genomic position of the genes
## Input:   data = data matrix; rows -> genes_hgnc
##          gencode = gencode object
genome_sort <- function(data, gencode){
    # detect organism
    if(str_detect(gencode$gene_id[1], "ENSG")){n_autosome=22}else if(str_detect(gencode$gene_id[1], "ENSMUSG")){n_autosome=19}
    data <- as.matrix(data)
    genes.have <- intersect(row.names(data),gencode$gene_name)
    data <- data[genes.have,]
    gencode <- gencode[match(row.names(data),gencode$gene_name),]
    gencode$chr <- factor(gsub("chr","",gencode$seqid),levels=c(seq(1,n_autosome),"X","Y"))
    filter <- factor(c(seq(1,n_autosome),"X"),levels=c(seq(1,n_autosome),"X")) # Y rausschmeiÃŸen
    gencode <- gencode[which(gencode$chr %in% filter),]
    gencode <- gencode[order(gencode$chr,gencode$start),]
    gencode$ensembl <- sapply(gencode$gene_id,function(x) strsplit(x,"\\.")[[1]][1])
    data <- data[match(gencode$gene_name,row.names(data)),]
    chromosomeorder <- data.frame(chr=gencode$chr,
                                  start=gencode$start,
                                  end=gencode$end,
                                  ensembl=gencode$ensembl,
                                  hgnc=gencode$gene_name)
    assign("chromosomeorder", chromosomeorder, .GlobalEnv)
    return(data)
}

## Smoothing gene-specific CNA scores; rows -> genes_hgnc; cols -> cells
## Input:   data = data matrix with gene-specific CNA scores
##          chromosomeorder = data frame with chromosomal order generated by sort_genome
##          windowlength = vector with length of moving window
##          weight = spearman correlation values for genes in network
## Output:  data frame with relative CNA profiles from all input cells
cnvperchr.weight <- function(data, chromosomeorder, windowlength){
    chrs <- as.character(unique(chromosomeorder$chr))
    #browser()
    for (i in chrs) {
        chr_genes <- as.character(chromosomeorder$hgnc[chromosomeorder$chr==i])
        cat("Number of genes on chromosome:\n")
        print(i)
        print(length(chr_genes))
        if(length(chr_genes)<2){next}
        chr_data <- data[chr_genes,]
        #print(dim(data))
        chr_cnv <- cnv.filter.weight(chr_data, windowlength)
        if(i==chrs[1]){
            rCNA <- chr_cnv
        }
        else {
            rCNA <- rbind(rCNA,chr_cnv)
        }
    }
    return(rCNA)
}

## Smoothing gene-specific CNA scores over a single chromosome, including correlations
## Input:   matrix = data matrix with gene-specific CNA scores; rows -> genes_hgnc; cols -> cells
##          windowlength = vector with length of moving window
##          weight = spearman correlation values for genes in network
cnv.filter.weight <- function(matrix, windowlength){
    for(cell in 1:ncol(matrix)){
        data <- cnv.filter.helper(matrix[,cell], windowlength[cell])
        if(cell==1){
            mat <- data
        } else {
            mat <- cbind(mat,data)
        }
    }
    colnames(mat) <- colnames(matrix)
    row.names(mat) <- row.names(matrix)
    return(mat)
}

## Smoothing gene-specific CNA scores over a single chromosome in a single cell, including correlations
## Input:   matrix = vector of gene-specific CNA-scores for one chromosome in a single cell
##          windowlength = length of moving window
##          weight = spearman correlation values for genes in network
cnv.filter.helper <- function(vector, windowlength){
    result <- numeric()

        if(length(vector) < windowlength){
            result <- rep(NA, length(vector))
        } else {
            for (i in c(1:length(vector))) {
                length <- round((windowlength-1)/2,0)
                lowerend <- ifelse((i-length)<1,1,(i-length))
                upperend <- ifelse((i+length)>length(vector),length(vector),(i+length))
                chr_vec <- vector[lowerend:upperend]
                result[i] <- mean(chr_vec)
                if(i < length){
                    result[i]	<- NA
                }
                if(i > length(vector)-length){
                    result[i]	<- NA
                }
            }
        }

    return(result)
}

